### DISCUSSION TITLE
Ref<T,PK> with Non-Standard id Property Name Causes Typescript Error

### DISCUSSION THREAD

**Discussion by vxm5091** [mikro-orm/mikro-orm#3964] [Q&A] [answered]
**Describe the bug**
Sorry in advance if this question is better suited for stackoverflow -- not sure if this is a bug or I'm misunderstanding the documentation. 

```typescript
people.entity.ts

@Entity({ schema: 'private', customRepository: () => PeopleRepository })
export class People {
  [OptionalProps]?: 'createdAt' | 'personId' | 'updatedAt';

  @PrimaryKey({
    defaultRaw: `next_id_varchar()`,
    fieldName: 'person_id',
  })
  personId!: string;

}
```

```typescript
peopleCheckinParks.entity.ts

@Entity({
  schema: 'private',
  customRepository: () => PeopleCheckinParksRepository,
})
export class PeopleCheckinParks {
  [OptionalProps]?: 'checkinAt' | 'checkinId';

  @PrimaryKey({ columnType: 'int8', defaultRaw: `next_id()` })
  checkinId!: string;

  @ManyToOne({
    entity: () => People,
    fieldName: 'person_id',
    onUpdateIntegrity: 'cascade',
    onDelete: 'cascade',
  })
  person!: Ref<People>;
```

When using `Ref<T>` or `Ref<T, PK>` on the `@ManyToOne` relationship, I'm noticing a few things that I'm having trouble wrapping my head around. 

1. Typescript won't let me query `PeopleCheckinParks` by `person` as a string. 
```typescript
await this.em.findOne(PeopleCheckinParks, { person: 'idstring' })
```
`Type 'string' is not assignable to type 'FilterQuery<People> | FilterValue<People>'`

Fixes I tried:
- Added this line to `People` : `[PrimaryKeyProp]!: 'personId'`
- Added the second generic type argument to `PeopleCheckinParks` : `person!: Ref<People, 'personId'>`

However, when I change the `personId` property to `id` or `_id` in `People`, the error goes away. Makes sense because `Loaded` checks for `id / _id / uuid`, but otherwise it doesn't seem to infer the custom `PK` string. 

Using this query as an example again for (2) and (3)
```typescript
const checkin = await this.em.findOne(PeopleCheckinParks, {
      person: 'idstring',
    });
```

2. Adding either the `PrimaryKeyProp` or the second generic argument doesn't statically expose that custom property on the `Loaded<PeopleCheckinParks, never>` query result from above. 

3. When the second generic argument in the reference type matches the property name in `People`, this is the type of `checkin.person`

`PeopleCheckinParks.person: {
    personId: string;
} & Reference<People>`

```typescript
people.entity.ts
export class People {
...
  @PrimaryKey({
    defaultRaw: `next_id_varchar()`,
    fieldName: 'person_id',
  })
 personId!: string; // <----- MATCH
```

```typescript
peopleCheckinParks.entity.ts

export class PeopleCheckinParks {
...

  @ManyToOne({
    entity: () => People,
    fieldName: 'person_id',
    onUpdateIntegrity: 'cascade',
    onDelete: 'cascade',
  })
  person!: Ref<People, 'personId>; // <----- MATCH
```

I think this makes sense - it's the intersection between the reference (not loaded) and the FK field. 

But when the second argument is still included but doesn't match -- this is resulting type:
`PeopleCheckinParks.person: {
    [OptionalProps]: "personId" | "createdAt" | "updatedAt";
    [EntityRepositoryType]: PeopleRepository;
    personId: string;
    username: string;
    personName: string;
    ... 20 more ...;
    buddies: Collection<...>;
} & Reference<...>`

TLDR;
- Seems like the `Ref` typing requires either the second generic argument to match the referenced PK property or not to be included at all in order to accurately reflect the `Loaded` type 
- Regardless of the above, the primary key property on the target entity class needs to be one of `id` / `_id` / `uuid` in order to query that field by value, even with `[PrimaryKeyProp]` included. 

Thanks in advance, hope this was helpful. 

**Versions**
typescript | 4.9.4 
mikro-orm | 5.6.7
postgres | 5.6.7

**Comment by B4nan** [comment, marked as answer]
You will need both `PrimaryKeyProp` and `PrimaryKeyType` on the entity class. The second type argument in `Ref` defaults to `PrimaryProperty` type, which [respects `PrimaryKeyType`](https://github.com/mikro-orm/mikro-orm/blob/60fee5350acc8dd32dc8863dc8efb99c6b451afd/packages/core/src/typings.ts#L43).

I am not sure if that combination works for you, as you don't have either of those in your entity defintion example? It should.

Btw unless you are using ts-morph, you need to add `ref: true` to the options if you want to use the reference wrapper.

**Reply by vxm5091** [reply]
> Btw unless you are using ts-morph, you need to add `ref: true` to the options if you want to use the reference wrapper.

I am using ts-morph so this seems to work. The one area I'm having an issue is when including an Embeddable field. Haven't gotten around to debugging that yet so for now including it as `@Embedded({ entity: () => AddressEmbeddable, nullable: true, object: true })`

Thanks for the heads up!

**Comment by vxm5091** [comment]
Thanks a lot for the clarification @B4nan! This works / makes a lot more sense. 

I didn't fully appreciate the distinction between `PrimaryKeyType` and `PrimaryKeyProp` in that they both check for one of `id` `_id` `uuid`. Thought `PrimaryKeyType` was only needed in the case of a composite key. Going through the docs again realizing I missed this helpful tidbit - `PrimaryKeyType is not needed when your entity has single scalar primary key under one of following property names: id: number | string | bigint, _id: any or uuid: string`. 

Confirmed, just me misunderstanding :)

**Reply by B4nan** [reply]
I will try to unify this in v6, the `PrimaryKeyProp` should be enough for both, iirc I tried to do it once and I failed to find the right way, but let's hope both TS capabilities as well as mine advanced and it will work this time :]

I was also thinking about a different approach, as something like an identity type might work as well:

```ts
@PrimaryKey({
  defaultRaw: `next_id_varchar()`,
  fieldName: 'person_id',
})
personId!: Primary<string>;
```

This could also allow the distinction of PK types on TS level (User PK is not the same as Company PK, even though they are both integers).
